### ISSUE TITLE
The return of Issue #6: Cannot read property 'displayName' of undefined

### ISSUE THREAD

**Issue by yormi** [bruderstein/unexpected-react#9]
Hola !

I encountered a problem that seems very similar to https://github.com/bruderstein/unexpected-react/issues/6

I've been trying for the whole afternoon to figure out a way to reproduce it with a simplified example and I finaaaaaallyyyy found what cause the problem(I can now continue my day in peace).

First, here's the error:
>Uncaught TypeError: Cannot read property 'displayName' of undefined
>      at ReactElementAdapter.getName (node_modules/unexpected-htmllike-jsx-adapter/lib/ReactElementAdapter.js:79:32)
>      at convertToDiff (node_modules/unexpected-htmllike/lib/convertToDiff.js:29:23)
>      at node_modules/unexpected-htmllike/lib/diff.js:538:69
>      at Array.forEach (native)
>      at node_modules/unexpected-htmllike/lib/diff.js:531:21
>      at node_modules/array-changes-async/lib/arrayChanges.js:224:17
>      at setEqual (node_modules/array-changes-async/lib/arrayChanges.js:199:28)
>      at node_modules/array-changes-async/lib/arrayChanges.js:212:25
>      at node_modules/unexpected-htmllike/lib/diff.js:505:24
>      at tryCatcher (node_modules/unexpected/node_modules/bluebird/js/main/util.js:26:23)
>      at Promise._settlePromiseFromHandler (node_modules/unexpected/node_modules/bluebird/js/main/promise.js:503:31)
>      at Promise._settlePromiseAt (node_modules/unexpected/node_modules/bluebird/js/main/promise.js:577:18)
>      at Async._drainQueue (node_modules/unexpected/node_modules/bluebird/js/main/async.js:128:12)
>      at Async._drainQueues (node_modules/unexpected/node_modules/bluebird/js/main/async.js:133:10)
>      at Async.drainQueues (node_modules/unexpected/node_modules/bluebird/js/main/async.js:15:14)
>      at node_modules/unexpected/lib/workQueue.js:8:13
>      at Array.forEach (native)
>      at Object.workQueue.drain (node_modules/unexpected/lib/workQueue.js:7:20)
>      at Immediate._onImmediate (node_modules/unexpected/lib/workQueue.js:17:19)

Here's how to reproduce it (I copy-pasted the required setup fromthe doc):

```javascript
class LiElement extends React.Component {
  render () {
    return <li>{this.props.name}</li>
  }
}

LiElement.propTypes = {
  name: React.PropTypes.string
}

class UlElement extends React.Component {
  render () {
    // if I change the 'LiElement' for a 'li' directly, no problem
    return (
      <ul>
        { this.props.items.map(i => <LiElement key={i.id} name={i.name} />) }
      </ul>
    )
  }
}

UlElement.propTypes = {
  items: React.PropTypes.array
}

const ReactTestUtils = require('react-addons-test-utils')

describe('Test', function () {
  const items = [
    { id: 0, name: 'Banana' },
    { id: 1, name: 'Chocolate' },
    { id: 2, name: 'Mustard' }
  ]

  it('should render liElement', function () {
    let component = ReactTestUtils.renderIntoDocument(
      <UlElement items={items} />
    )

    return expect(component, 'to have rendered with all children', (
      <ul>
        <li>{items[0].name}</li>
        <li>{items[2].name}</li>
        <li>{items[1].name}</li>
      </ul>
    )) // displayName if not in the right order, no error otherwise
  })
})
```

**Comment by bruderstein** [comment]
Ok, I can reproduce it - thanks for a great bug report.

I'm investigating...

**Comment by bruderstein** [comment]
Wow. This a great bug. I've found the issue. I should have it fixed today or latest tomorrow (the fix is pretty simple, but I want to add some tests around it).

**Comment by bruderstein** [comment]
@yormi I've just published 0.5.2, which forces the update of unexpected-htmllike, which fixes this issue.  Thanks again for detailed bug report - I've added it in as a regression test.

**Comment by yormi** [comment]
Awesome man ! You are a really quick problem solver :) These days, I'm in holidays but I'll upgrade when I come back. Thank you for your great work ! :D

**Comment by RusAlex** [comment]
Thank you very much. Also met the bug.

**Comment by dlindahl** [comment]
I also ran into this bug when compiling with the `transform-react-constant-elements` plugin from [`babel-react-optimize`](https://github.com/thejameskyle/babel-react-optimize). I was able to workaround the issue by using the plugins explicitly (and omitting `transform-react-constant-elements` rather than using the Babel preset itself.

It'd be great if `unexpected-html` could not blow up when `displayName` is not available as I'd like to be able to test my components as they appear in production, but I realize this may not be a the common path and am OK moving forward with my workaround. Wanted to share my experience in case anyone else has the same setup as I do.
